### DISCUSSION TITLE
Adding version numbers to input files

### DISCUSSION THREAD

**Discussion by ECB2020** [Open-SES/OpenSES#21] [Ideas]
# Introduction

Hi all,

In the OpenSES meeting at the Ventilation and Fires in Tunnels conference in Brighton last month I offered to port into OpenSES a feature that the group may find useful.

I made a start on porting it, then realized that I need some input first if I'm not to start railroading OpenSES down the path that suits me best.  So here goes.

First, I'll describe the proposal in detail so that those who were not at the meeting can comment.

Second, I have a few questions for the group.

# Proposed code change

I propose to read a line after form 1B in the input file (before form 1C).

Once the line after form 1B is read, the program will look for the text "OpenSES" followed by a number. An example of the top of such an input file is as follows:

```
This is a line of comment at the top of an input file.
This is a second line of comment at the top of an input file.
This is the last line of comment and marks the end of form 1A.
17.       10        2022                                                         Form 1B
   OpenSES  4.9
1         1         1         1         0         4         0         0          Form 1C
```

If the text on the line is "OpenSES 4.2", then OpenSES expects all the input to be compatible with SES v4.1 and calculates in the same way as SES v4.1.

If the text on the line is "OpenSES 4.9", then OpenSES does something else.  It might:

- calculate in a different way (such as derating jet fans according to local air density).
- expect additional numbers in existing forms (such as segment fixed pressure rises),
- allow additional forms to be read (such as forms that define Saccardo nozzles),
- correct known bugs/undesirable behaviours that occur in v4.1.

We don't have to decide what that "something else" is yet, those are just possibilities.

The next time a new feature is added, the version number goes up by 0.1.

You may be wondering why we don't just have "4.9" on a line of its own. It's because I reckon that having the text "OpenSES" near the top of input is a useful marker that will let users distinguish between input files that are to be run in OpenSES and input files for v4.1/SVS.  I've also found that having bare numbers in input files can be a problem for users learning to use the software.

# Some implementation details
- The first 80 characters will be read from the line (like the other forms).  If you have 80 spaces before "OpenSES" it won't see it.
- The word "OpenSES" is not case-sensitive.
- The entries on the line are not fixed-format. You can have spaces before "OpenSES" and between "OpenSES" and the number.
- Incorrectly formed entries raise one of five errors in the same general style as the existing SES errors, as follows:
> 1. The line isn't present (the error complains about possibly reading form 1C).
> 2. The word "OpenSES" is present but there is no version number.
> 3. The word "OpenSES" is present but there is garbage instead of a version number.
> 4. The version number is too low (less than 4.2).
> 5. The version number is too high (higher than whatever the maximum version number this executable can handle).
- Although the version number in the input file is a floating-point number, internally that number is multiplied by 1000 and stored in integer variable "INPVER" in the common block in DSHARE.  I have had so many issues with comparing floating-point numbers in the past that I prefer to do integer comparisons wherever possible - it removes a whole class of comparison errors.

## Notes about  coding
- I usually write in Fortran 95 and never in Fortran IV (no arithmetic IFs, as few GOTOs as possible).
- I didn't use the existing SESVER variable because that's the version number of the executable, not the version number of the input file.
- When you need to interrogate the version number to decide how a feature should behave you can use code like:
```
  if (inpver .gt. 4200) then
     <do stuff, e.g. derate jet fans for local temperature>
  else
     <do other stuff, e.g. don't derate jet fans for local temperature>
  end if
```

# Questions

### 1st question:
The OpenSES paper at the ISAVFT describes a "Blessed OpenSES Repository". Is the code that I just git cloned from https://github.com/Open-SES/OpenSES the Blessed version?
If not, could one of the founders publish a blessed version so I can clone that?

### 2nd question:
Do you want to use 4.1, 4.2, 4.3 etc. for the input file version numbers or some other numbering scheme (e.g. 1, 2, 3)?

I prefer having the input files given versions like 4.1, 4.2, 4.3 etc. rather than 1, 2, 3 because it may concentrate the users' minds: "I'm writing a version 4.1 file so I shouldn't include extra features" or "I'm writing a version 4.7 input file so I can use Saccardo nozzles."

If there are no strong views I will start from 4.2.  4.2 will behave like 4.1.

### 3rd question:
Does anyone have an existing error numbering scheme?

If not then I will start numbering error messages from 1001. The routines that read the new line use errors 1001 to 1005.

### 4th question:
is there an experienced Fortran user willing to volunteer to review a set of changes to the COMMON block files (DSHARE and DSHRHS)?

I pretty much insist on using "implicit none" in Fortran (the result of a lot of bad experiences when I was programming in my twenties). In order to be able to use "implicit none" in new (.f95) routines, I had to heavily modify the two common block files DSHARE and DSHRHS to add explicit declarations of all the variables (about half of them used to be implicitly declared by their COMMON statements).

This is difficult to get right first time. But I've developed a process for refactoring old code to catch mismatches, as follows.

 - I wrote new versions of DSHARE and DSHRHS with everthing declared.
 - I made a short file that imports the old versions of DSHARE and DSHRHS.
 - I made a similar file that imports the new versions.
 - I compiled both of these with "gfortran -fdump-parse-tree", which tells the compiler to list the size and type of every variable and state whether it was implicitly or explicitly declared.
 - I compared the two transcripts and dealt with mismatches in variable type and array size that I'd got wrong. After a bit of rinse and repeat I ended up with transcripts in which the only differences were the implicit or explicit declaration.
 - I generated a diff of these two transcripts and the only entries in it were the implicit or explicit declaration.

I *think* this process led me to versions of DSHARE and DSHRHS that declare the same variables as the old code but which are suitable for use in new Fortran files that use "implicit none".

I would appreciate it if someone else would volunteer to review the records of this process and see if they can find a difference that I missed. The ideal volunteer would be someone who (like me) has fallen into a lot of elephant traps when mixing old-style Fortran with new-style Fortran.  Any takers?

**Comment by iamchermac** [comment]
It's a 'problem' I have thought about a few times over the years. Especially when thinking about how to internally tag input files to better work with scripts and applications I have developed.

I agree with the approach in principle, except for the addition of a line between Forms 1B and 1C. I've always thought that something like this should be implemented using entries on the Form 1B line to aid backwards compatibility.

SES 4.1 should still identify Form 1B for a line that's '17.       6.        2022.     1.        0.' has I've had past projects teach me to not have number formats at the beginning of Form 1A comments. In my example 1 could be for a representative 'computing version' (could be a decimal and then internally converted to an integer as suggested) and 0 is IP (1 could be SI). Backwards compatibility with as minimal intermediate conversions is ideal.

I understand the point with regards to text and user uptake. I agree to some extent. But I think that particular user uptake will crave something along the lines of an 'input manager'.

Regarding blessed source code: this should be the main branch. Development should take place on user created branches and once agreed, desired changes can be rebased.

Regarding error codes, do you mean the existing schema, or a new one? I think we have the table of existing error codes somewhere for reference. It's within the package of documentation that needs to be added to the site.

I'm currently in the process of retraining myself for Fortran. But once better up to speed, I hope to lend a hand with checking in-depth code.

**Reply by ECB2020** [reply]
Hi all,

I've worked up a modified version in a fork (just pushed it from offline to https://github.com/ECB2020/OpenSES-inp_ver).  Let me know if it isn't visible (I hardly ever use git, and pushing and pulling stuff from the command line is new to me).

It's a compromise between my suggestions and Chermac's: I took out out the separate line and the word "OpenSES" (which would have broken back compatibility with v4.1, because 4.1 reads eight numbers for form 1B) and put the version number into characters 31 to 40 of form 1B.

I didn't add a switch to choose between US and SI input, as I don't think you'll be ready to write US input files for some time.  It would be easy to add one in a later version of OpenSES in characters 41 to 50.

I've added some test files and README.md files to guide those of you wanting to check it over for inclusion and understand some of  the other changes that I've made, such as the ones to the makefile and FILESUBS.FOR.

Happy to answer any questions that come up during the review.

**Reply by NeverGray** [reply]
Hi ECB2020,
We received your request. We will discuss the best process to review your changes. Give us a week or two for an official response. 
Justin

**Comment by ECB2020** [comment]
Hi all,

I've finally managed to wrangle git enough to upload a version of my 'OpenSES-inp_ver' branch of that's based on the 'Develop' branch of OpenSES instead of the main branch.  The combined code now writes JSON files and looks for a version number in the input file.  The verification folder is a merged version of inp_ver and Develop.

A full description of the changes can be found in the README.md file of the 'source' folder.  The changes are mostly the same as the November version, but there is one extra, as I had to figure out how to cope with the string SESVER containing text as well as numbers.

Happy to clarify if anything is unclear.
